ddl/column.go

```diff
@@ -93,57 +99,116 @@ func (d *ddl) createColumnInfo(tblInfo *model.TableInfo, colInfo *model.ColumnIn
+			if err != nil {
+				job.State = model.JobCancelled
+				return nil, errors.Trace(err)
+			}
+		}
+		columnInfos = append(columnInfos, colInfo)
+	}
+	return columnInfos, nil
+}
+
+func (d *ddl) onAddColumns(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 	schemaID := job.SchemaID
 	tblInfo, err := getTableInfo(t, job, schemaID)
 	if err != nil {
 		return ver, errors.Trace(err)
 	}
 
-	col := &model.ColumnInfo{}
-	pos := &ast.ColumnPosition{}
-	offset := 0
-	err = job.DecodeArgs(col, pos, &offset)
-	if err != nil {
+	columns := []*model.ColumnInfo{}
+	positions := []*ast.ColumnPosition{}
+	lastColOffset := 0
+	if err = job.DecodeArgs(&columns, &positions, &lastColOffset); err != nil {
 		job.State = model.JobCancelled
 		return ver, errors.Trace(err)
 	}
+	// First run of this job, get the offset of last column.
+	if lastColOffset == 0 {
+		lastColOffset = len(tblInfo.Columns)
+	}
 
-	columnInfo := findCol(tblInfo.Columns, col.Name.L)
-	if columnInfo != nil {
-		if columnInfo.State == model.StatePublic {
-			// We already have a column with the same column name.
-			job.State = model.JobCancelled
-			return ver, infoschema.ErrColumnExists.GenByArgs(col.Name)
-		}
-	} else {
-		columnInfo, offset, err = d.createColumnInfo(tblInfo, col, pos)
-		if err != nil {
-			job.State = model.JobCancelled
-			return ver, errors.Trace(err)
-		}
-		// Set offset arg to job.
-		if offset != 0 {
-			job.Args = []interface{}{columnInfo, pos, offset}
-		}
+	if err = checkHasDuplicateColumnName(columns); err != nil {
+		job.State = model.JobCancelled
+		return ver, errors.Trace(err)
 	}
 
-	originalState := columnInfo.State
-	switch columnInfo.State {
+	columnsInfo, err := d.getColumnInfos(columns, positions, job, tblInfo, lastColOffset)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	// Check whether all columns have the same state.
+	originalState := columnsInfo[0].State
+	if !checkColumnsStateConsistency(columnsInfo) {
+		job.State = model.JobCancelled
+		return ver, ErrInvalidColumnState.Gen("unconsistency of all columns state: %v", originalState)
+	}
+
+	job.Args = []interface{}{columnsInfo, positions, lastColOffset}
+	switch columnsInfo[0].State {
 	case model.StateNone:
-		// none -> delete only
 		job.SchemaState = model.StateDeleteOnly
-		columnInfo.State = model.StateDeleteOnly
+		changeAllColumnsState(columnsInfo, model.StateDeleteOnly)
 		ver, err = updateTableInfo(t, job, tblInfo, originalState)
 	case model.StateDeleteOnly:
 		// delete only -> write only
 		job.SchemaState = model.StateWriteOnly
-		columnInfo.State = model.StateWriteOnly
+		changeAllColumnsState(columnsInfo, model.StateWriteOnly)
 		ver, err = updateTableInfo(t, job, tblInfo, originalState)
 	case model.StateWriteOnly:
 		// write only -> reorganization
 		job.SchemaState = model.StateWriteReorganization
-		columnInfo.State = model.StateWriteReorganization
+		changeAllColumnsState(columnsInfo, model.StateWriteReorganization)
```

Comment:
I think `setColumnState` is better than `changeAllColumnsState`. what do you think?